File: VerbumKit/Sources/Features/Scripture/ChapterReaderFeature.swift
import Clients
import ComposableArchitecture
import Foundation
import Models

/// One chapter on screen: loads it, lets the reader select verses, steps to
/// the neighbouring chapters. Owns no navigation; the parent decides where
/// the reader goes.
@Reducer
public struct ChapterReaderFeature {
    @ObservableState
    public struct State: Equatable {
        public var reference: PassageReference
        @Shared(.readingActivity) var readingActivity
        public var content: Content = .idle
        public var selectedVerses: Set<Int> = []
        public var requestedVerses: ClosedRange<Int>?
        @Shared(.readerTextScale) public var textScale
        @Shared(.lastRead) public var lastRead
        @Shared(.readingMode) public var readingMode
        @Shared(.readerFocusMode) public var focusMode
        @Presents public var study: VerseStudyFeature.State?
        public var chapters: [String: [BiblePassage]] = [:]
        public var chapterErrors: [String: ReaderError] = [:]
        public var loadingChapters: Set<String> = []
        public var contexts: [String: PassageContext] = [:]
        public var mentions: [String: [StudyTextSegment]] = [:]
        public var annotations: [String: ReaderAnnotation] = [:]
        public var annotationLoadFailed = false
        public var flow: [PassageReference] = []
        public var history: [Visit] = []
        public var scrollOffsets: [String: Double] = [:]
        public var restoreOffset: Double?
        public var navigationRevision = 0
        public struct Visit: Equatable, Sendable {
            public var reference: PassageReference
            public var flow: [PassageReference]
            public var offset: Double
        }

        public init(reference: PassageReference) {
            self.reference = PassageReference(bookId: reference.bookId, chapter: reference.chapter)
            self.requestedVerses = reference.verses
            self.flow = [self.reference]
        }

        public var book: BibleBook? { BibleBook.book(id: reference.bookId) }
        public var title: String { reference.formatted }

        public var canGoToNextChapter: Bool { ChapterNavigation.next(after: reference) != nil }
        public var canGoToPreviousChapter: Bool { ChapterNavigation.previous(before: reference) != nil }

        /// `John 3:16-18, 21` while verses are selected.
        public var selectionCitation: String? {
            SelectionFormatter.format(bookId: reference.bookId, chapter: reference.chapter, verses: selectedVerses)
        }

        public var selectedText: String? {
            guard case .loaded(let verses) = content, !selectedVerses.isEmpty else { return nil }
            return verses
                .filter { selectedVerses.contains($0.verseStart) }
                .map(\.text)
                .joined(separator: " ")
        }
    }

    public enum Content: Equatable, Sendable {
        case idle
        case loading
        case loaded([BiblePassage])
        case failed(ReaderError)
    }

    public enum Action: Equatable {
        case task, recordReading
        case ensureChapter(PassageReference)
        case cachedChapter(PassageReference, Result<[BiblePassage], ReaderError>)
        case contextResponse(PassageReference, PassageContext?)
        case annotationsResponse([ReaderAnnotation]), annotationsFailed
        case study(PresentationAction<VerseStudyFeature.Action>)
        case studyVerse(PassageReference, Int, [String])
        case appendChapter
        case chapterVisible(PassageReference)
        case visit(PassageReference), backToReading, focusToggled, readingModeChanged
        case scrollOffsetChanged(String, Double)
        case retryTapped
        case chapterResponse(Result<[BiblePassage], ReaderError>)
        case verseTapped(Int)
        case clearSelectionTapped
        case copySelectionTapped
        case nextChapterTapped
        case previousChapterTapped
        /// Parent-driven jump (book picker). Reloads.
        case go(to: PassageReference)
        case listenTapped
        case talkTapped
        case contextTapped
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case listen(PassageReference)
            /// Start a spoken conversation about this chapter.
            case talk(PassageReference)
            case openContext(PassageReference)
        }
    }

    @Dependency(\.date.now) var now
    @Dependency(\.calendar) var calendar
    @Dependency(\.bibleClient) var bibleClient
    @Dependency(\.pasteboard) var pasteboard
    @Dependency(\.contextClient) var contextClient
    @Dependency(\.readerAnnotations) var readerAnnotations

    public init() {}

    public var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .recordReading:
                guard state.chapters[ReaderCanon.key(state.reference)]?.isEmpty == false else { return .none }
                state.$readingActivity.withLock { $0.record(state.reference, at: now, calendar: calendar) }
                return .none
            case .task:
                let annotationEffect: Effect<Action> = .run { [readerAnnotations] send in
                    do { await send(.annotationsResponse(try await readerAnnotations.load())) }
                    catch { await send(.annotationsFailed) }
                }.cancellable(id: CancelID.annotations, cancelInFlight: true)
                if case .loaded = state.content { return annotationEffect }
                return .merge(load(&state), annotationEffect)
            case .annotationsResponse(let values):
                state.annotations = Dictionary(values.map { ($0.id, $0) }, uniquingKeysWith: { _, newest in newest })
                state.annotationLoadFailed = false; return .none
            case .annotationsFailed: state.annotationLoadFailed = true; return .none
            case .ensureChapter(let reference):
                let key = ReaderCanon.key(reference)
                guard state.chapters[key] == nil, !state.loadingChapters.contains(key) else { return .none }
                state.loadingChapters.insert(key)
                return .run { [bibleClient] send in
                    do { await send(.cachedChapter(reference, .success(try await bibleClient.chapter(bookId: reference.bookId, chapter: reference.chapter)))) }
                    catch is CancellationError {} catch { await send(.cachedChapter(reference, .failure(ReaderError(error)))) }
                }.cancellable(id: key, cancelInFlight: true)
            case .cachedChapter(let reference, let result):
                let key = ReaderCanon.key(reference); state.loadingChapters.remove(key)
                switch result {
                case .success(let verses):
                    state.chapters[key] = verses; state.chapterErrors[key] = nil
                    if state.reference == reference { return .send(.chapterResponse(.success(verses))) }
                case .failure(let error):
                    state.chapterErrors[key] = error
                    if state.reference == reference { state.content = .failed(error) }
                }
                return .none
            case .contextResponse(let reference, let context):
                if let context {
                    let key = ReaderCanon.key(reference)
                    state.contexts[key] = context
                    for verse in state.chapters[key] ?? [] {
                        state.mentions["\(verse.bookId).\(verse.chapter).\(verse.verseStart)"] = ReaderEntityLinker.segments(verse.text, entities: context.entities)
                    }
                }
                return .none
            case .studyVerse(let reference, let number, let ids):
                guard let verse = state.chapters[ReaderCanon.key(reference)]?.first(where: { $0.verseStart == number }) else { return .none }
                let key = "\(reference.bookId).\(reference.chapter).\(number)"
                let context = state.contexts[ReaderCanon.key(reference)]
                state.study = VerseStudyFeature.State(passage: verse, annotation: state.annotations[key], context: context,
                    candidates: context?.entities.filter { ids.contains($0.id) } ?? [])
                return .none
            case .study(.presented(.delegate(.annotationSaved(let annotation)))):
                state.annotations[annotation.id] = annotation; return .none
            case .study(.presented(.delegate(.openPassage(let reference)))):
                state.study = nil; return .send(.visit(reference))
            case .study(.presented(.delegate(.close))): state.study = nil; return .none
            case .study: return .none
            case .visit(let reference):
                state.history.append(State.Visit(reference: state.reference, flow: state.flow,
                    offset: state.scrollOffsets[state.readingMode == .continuous ? "flow" : ReaderCanon.key(state.reference)] ?? 0))
                if state.history.count > 32 { state.history.removeFirst() }
                return jump(&state, to: reference)
            case .backToReading:
                guard let visit = state.history.popLast() else { return .none }
                let effect = jump(&state, to: visit.reference)
                state.flow = visit.flow; state.restoreOffset = visit.offset
                return effect
            case .scrollOffsetChanged(let key, let offset): state.scrollOffsets[key] = offset; return .none
            case .focusToggled: state.$focusMode.withLock { $0.toggle() }; return .none
            case .readingModeChanged:
                state.flow = [state.reference]; state.navigationRevision += 1
                state.restoreOffset = 0; state.requestedVerses = nil
                return .none
            case .appendChapter:
                guard let last = state.flow.last, state.chapters[ReaderCanon.key(last)] != nil,
                      let next = ChapterNavigation.next(after: last) else { return .none }
                state.flow.append(next)
                return .send(.ensureChapter(next))
            case .chapterVisible(let reference):
                guard state.readingMode == .continuous, state.reference != reference else { return .none }
                state.reference = reference
                if let verses = state.chapters[ReaderCanon.key(reference)] { state.content = .loaded(verses) }
                state.$lastRead.withLock { $0 = reference }
                return contextEffect(reference)

            case .retryTapped:
                return load(&state)

            case .chapterResponse(.success(let verses)):
                if let first = verses.first, first.bookId != state.reference.bookId || first.chapter != state.reference.chapter { return .none }
                state.loadingChapters.remove(ReaderCanon.key(state.reference))
                state.content = .loaded(verses)
                state.chapters[ReaderCanon.key(state.reference)] = verses
                if let requested = state.requestedVerses {
                    state.selectedVerses = Set(verses.map(\.verseStart).filter { requested.contains($0) })
                }
                let reference = state.reference
                state.$lastRead.withLock { $0 = reference }
                let neighbors = [ChapterNavigation.previous(before: reference), ChapterNavigation.next(after: reference)].compactMap { $0 }
                return .merge([contextEffect(reference)] + neighbors.map { .send(.ensureChapter($0)) })

            case .chapterResponse(.failure(let error)):
                state.content = .failed(error)
                state.loadingChapters.remove(ReaderCanon.key(state.reference))
                state.chapterErrors[ReaderCanon.key(state.reference)] = error
                return .none

            case .verseTapped(let verse):
                return .send(.studyVerse(state.reference, verse, []))

            case .clearSelectionTapped:
                state.selectedVerses = []
                return .none

            case .copySelectionTapped:
                guard let citation = state.selectionCitation, let text = state.selectedText else { return .none }
                return .run { [pasteboard] _ in pasteboard.copy(text: "\(text)\n— \(citation)") }

            case .nextChapterTapped:
                guard let next = ChapterNavigation.next(after: state.reference) else { return .none }
                return jump(&state, to: next)

            case .previousChapterTapped:
                guard let previous = ChapterNavigation.previous(before: state.reference) else { return .none }
                return jump(&state, to: previous)

            case .go(let reference):
                return jump(&state, to: reference)

            case .listenTapped:
                return .send(.delegate(.listen(state.reference)))

            case .talkTapped:
                return .send(.delegate(.talk(state.reference)))

            case .contextTapped:
                return .send(.delegate(.openContext(state.reference)))

            case .delegate:
                return .none
            }
        }.ifLet(\.$study, action: \.study) { VerseStudyFeature() }
    }

    private func contextEffect(_ reference: PassageReference) -> Effect<Action> {
        .run { [contextClient] send in
            let context = try? await contextClient.chapter(reference: reference)
            await send(.contextResponse(reference, context))
        }.cancellable(id: "context-" + ReaderCanon.key(reference), cancelInFlight: true)
    }

    private func jump(_ state: inout State, to reference: PassageReference) -> Effect<Action> {
        state.reference = PassageReference(bookId: reference.bookId, chapter: reference.chapter)
        state.requestedVerses = reference.verses
        state.selectedVerses = []
        state.flow = [state.reference]
        state.navigationRevision += 1
        state.restoreOffset = reference.verses == nil ? (state.readingMode == .continuous ? 0 : state.scrollOffsets[ReaderCanon.key(state.reference)] ?? 0) : nil
        if let verses = state.chapters[ReaderCanon.key(state.reference)] {
            state.content = .loaded(verses)
            state.$lastRead.withLock { $0 = state.reference }
            return .merge(contextEffect(state.reference), .send(.ensureChapter(ChapterNavigation.next(after: state.reference) ?? state.reference)), .send(.ensureChapter(ChapterNavigation.previous(before: state.reference) ?? state.reference)))
        }
        return load(&state)
    }

    private func load(_ state: inout State) -> Effect<Action> {
        state.content = .loading
        // Every response carries its chapter identity, including rapid navigation.
        // Shared per-chapter requests finish into the cache without changing another page.
        return .send(.ensureChapter(state.reference))
    }

    private enum CancelID { case load, annotations }
}

extension Result where Failure == any Error {
    fileprivate func mapError<E: Error>(_ transform: (any Error) -> E) -> Result<Success, E> {
        switch self {
        case .success(let value): .success(value)
        case .failure(let error): .failure(transform(error))
        }
    }
}

File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/ChapterReaderFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.*
import com.nexussoft.verbum.common.arch.*
import com.nexussoft.verbum.models.*
import kotlinx.coroutines.CancellationException

/** Native pages and the continuous flow share a single reader domain and bounded prefetch. */
object ChapterReaderFeature {
    data class Visit(val reference: PassageReference,val flow: List<PassageReference>,val index: Int,val offset: Int)
    data class Position(val index: Int=0,val offset: Int=0)
    data class State(
        val reference: PassageReference,
        val content: Content = Content.Idle,
        val selectedVerses: Set<Int> = emptySet(),
        val textScale: ReaderTextScale = ReaderTextScale.STANDARD,
        val requestedVerses: IntRange? = reference.verses,
        val readingMode: ReadingMode = ReadingMode.PAGES,
        val focusMode: Boolean=false,
        val study: VerseStudyFeature.State?=null,
        val chapters: Map<String,List<BiblePassage>> = emptyMap(),
        val chapterErrors: Map<String,ReaderError> = emptyMap(),
        val loadingChapters: Set<String> = emptySet(),
        val contexts: Map<String,PassageContext> = emptyMap(),
        val mentions: Map<String,List<StudyTextSegment>> = emptyMap(),
        val annotations: Map<String,ReaderAnnotation> = emptyMap(),
        val annotationLoadFailed: Boolean=false,
        val flow: List<PassageReference> = listOf(PassageReference(reference.bookId,reference.chapter)),
        val history: List<Visit> = emptyList(),
        val positions: Map<String,Position> = emptyMap(),
        val restorePosition: Position?=null,
        val navigationRevision: Int=0,
    ) {
        constructor(reference: PassageReference,textScale: ReaderTextScale):this(PassageReference(reference.bookId,reference.chapter),Content.Idle,emptySet(),textScale,reference.verses)
        val book: BibleBook? get()=BibleBook.book(reference.bookId)
        val title: String get()=reference.formatted
        val canGoToNextChapter get()=ChapterNavigation.next(reference)!=null
        val canGoToPreviousChapter get()=ChapterNavigation.previous(reference)!=null
        val selectionCitation: String? get()=SelectionFormatter.format(reference.bookId,reference.chapter,selectedVerses)
        val selectedText: String? get()=(content as? Content.Loaded)?.verses?.filter { it.verseStart in selectedVerses }?.takeIf { it.isNotEmpty() }?.joinToString(" ") { it.text }
    }
    sealed interface Content {
        data object Idle:Content;data object Loading:Content
        data class Loaded(val verses: List<BiblePassage>):Content
        data class Failed(val error: ReaderError):Content
    }
    sealed interface Action {
        data object Started:Action;data object RetryTapped:Action
        data class ChapterLoaded(val verses:List<BiblePassage>):Action
        data class ChapterFailed(val error:ReaderError):Action
        data class EnsureChapter(val reference:PassageReference):Action
        data class CachedChapter(val reference:PassageReference,val verses:List<BiblePassage>?,val error:ReaderError?=null):Action
        data class ContextLoaded(val reference:PassageReference,val context:PassageContext?):Action
        data class AnnotationsLoaded(val values:List<ReaderAnnotation>):Action
        data object AnnotationsFailed:Action
        data class PreferencesLoaded(val mode:ReadingMode,val focus:Boolean):Action
        data class Study(val action:VerseStudyFeature.Action):Action
        data object StudyDismissed:Action
        data class StudyVerse(val reference:PassageReference,val verse:Int,val ids:List<String> = emptyList()):Action
        data class VerseTapped(val verse:Int):Action
        data object ClearSelectionTapped:Action;data object CopySelectionTapped:Action
        data object NextChapterTapped:Action;data object PreviousChapterTapped:Action
        data class Go(val reference:PassageReference):Action
        data class VisitPassage(val reference:PassageReference):Action
        data object BackToReading:Action;data object AppendChapter:Action
        data class ChapterVisible(val reference:PassageReference):Action
        data class PositionChanged(val key:String,val index:Int,val offset:Int):Action
        data class TextScaleChanged(val scale:ReaderTextScale):Action
        data class ModeChanged(val mode:ReadingMode):Action
        data class FocusChanged(val enabled:Boolean):Action
        data object FocusToggled:Action
        data object ListenTapped:Action;data object TalkTapped:Action;data object ContextTapped:Action
        data class Delegate(val delegate:DelegateAction):Action
    }
    sealed interface DelegateAction {
        data class Listen(val reference:PassageReference):DelegateAction
        data class Talk(val reference:PassageReference):DelegateAction
        data class OpenContext(val reference:PassageReference):DelegateAction
    }
    // Requests are keyed by canonical chapter identity.
    const val LAST_READ_KEY="lastRead"
    const val MODE_KEY="readingMode"
    const val FOCUS_KEY="readerFocusMode"
    fun reducer(bibleClient:BibleClient,clipboard:ClipboardClient,preferences:PreferencesClient,
        contextClient:ContextClient = com.nexussoft.verbum.clients.fixtures.FixtureContextClient,
        graphClient:GraphClient = com.nexussoft.verbum.clients.fixtures.FixtureGraphClient,
        askClient:AskScriptureClient = AskScriptureClient { throw AskScriptureException.Unavailable },
        annotations:ReaderAnnotationsClient = PreferenceReaderAnnotationsClient(preferences),
        comparison:BibleComparisonClient = LiveBibleComparisonClient(),
    ):Reducer<State,Action> = combine(
        VerseStudyFeature.reducer(contextClient,graphClient,askClient,annotations,comparison).pullbackOptional(
            get={it.study},set={s,c->s.copy(study=c)},extractAction={(it as? Action.Study)?.action},embedAction={Action.Study(it)},
        ),
        Reducer {state,action ->
            fun context(reference:PassageReference):Effect<Action> = runEffect(id="context-${ReaderCanon.key(reference)}",cancelInFlight=true) {send->
                preferences.setString(LAST_READ_KEY,LastRead.encode(reference))
                val value=try {contextClient.chapter(reference)} catch(e:CancellationException){throw e} catch(e:Exception){null}
                send(Action.ContextLoaded(reference,value))
            }
            when(action) {
                Action.Started -> state.with(runEffect {send->
                    val mode=preferences.string(MODE_KEY)?.let { runCatching { ReadingMode.valueOf(it) }.getOrNull() } ?: ReadingMode.PAGES
                    send(Action.PreferencesLoaded(mode,preferences.string(FOCUS_KEY)=="true"))
                    try {send(Action.AnnotationsLoaded(annotations.load()))} catch(e:CancellationException){throw e} catch(e:Exception){send(Action.AnnotationsFailed)}
                    if(state.content !is Content.Loaded) send(Action.RetryTapped)
                })
                Action.RetryTapped -> load(state,bibleClient)
                is Action.PreferencesLoaded -> state.copy(readingMode=action.mode,focusMode=action.focus).only()
                is Action.AnnotationsLoaded -> state.copy(annotations=action.values.associateBy { it.id },annotationLoadFailed=false).only()
                Action.AnnotationsFailed -> state.copy(annotationLoadFailed=true).only()
                is Action.ChapterLoaded -> {
                    val first=action.verses.firstOrNull()
                    if(first!=null && (first.bookId!=state.reference.bookId || first.chapter!=state.reference.chapter)) return@Reducer state.only()
                    val key=ReaderCanon.key(state.reference)
                    state.copy(content=Content.Loaded(action.verses),chapters=state.chapters+(key to action.verses),loadingChapters=state.loadingChapters-key,
                        selectedVerses=state.requestedVerses?.let {range->action.verses.map {it.verseStart}.filter {it in range}.toSet()} ?: emptySet()).with(Effect.Merge(listOf(
                            context(state.reference),runEffect {send->
                                preferences.setString(LAST_READ_KEY,LastRead.encode(state.reference))
                                ChapterNavigation.previous(state.reference)?.let {send(Action.EnsureChapter(it))}
                                ChapterNavigation.next(state.reference)?.let {send(Action.EnsureChapter(it))}
                            },
                        )))
                }
                is Action.ChapterFailed -> state.copy(content=Content.Failed(action.error),loadingChapters=state.loadingChapters-ReaderCanon.key(state.reference),chapterErrors=state.chapterErrors+(ReaderCanon.key(state.reference) to action.error)).only()
                is Action.EnsureChapter -> {
                    val key=ReaderCanon.key(action.reference)
                    if(key in state.chapters || key in state.loadingChapters) state.only()
                    else state.copy(loadingChapters=state.loadingChapters+key,chapterErrors=state.chapterErrors-key).with(runEffect(id="chapter-$key",cancelInFlight=true) {send->
                        try {send(Action.CachedChapter(action.reference,bibleClient.chapter(action.reference.bookId,action.reference.chapter)))}
                        catch(e:CancellationException){throw e} catch(e:Exception){send(Action.CachedChapter(action.reference,null,ReaderError.from(e)))}
                    })
                }
                is Action.CachedChapter -> {
                    val key=ReaderCanon.key(action.reference)
                    if(action.verses!=null) state.copy(chapters=state.chapters+(key to action.verses),loadingChapters=state.loadingChapters-key,chapterErrors=state.chapterErrors-key,
                        content=if(state.reference==action.reference) Content.Loaded(action.verses) else state.content).with(if(state.reference==action.reference) Effect.Send(Action.ChapterLoaded(action.verses)) else Effect.None)
                    else state.copy(loadingChapters=state.loadingChapters-key,chapterErrors=state.chapterErrors+(key to (action.error ?: ReaderError.from(Exception()))),
                        content=if(state.reference==action.reference) Content.Failed(action.error ?: ReaderError.from(Exception())) else state.content).only()
                }
                is Action.ContextLoaded -> {
                    val value=action.context
                    if(value==null) state.only() else {
                        val key=ReaderCanon.key(action.reference)
                        val mentions=state.chapters[key].orEmpty().associate { "${it.bookId}.${it.chapter}.${it.verseStart}" to ReaderEntityLinker.segments(it.text,value.entities) }
                        state.copy(contexts=state.contexts+(key to value),mentions=state.mentions+mentions).only()
                    }
                }
                is Action.StudyVerse -> {
                    val verse=state.chapters[ReaderCanon.key(action.reference)]?.firstOrNull {it.verseStart==action.verse}
                    if(verse==null) state.only() else {
                        val value=state.contexts[ReaderCanon.key(action.reference)]
                        state.copy(study=VerseStudyFeature.State.initial(verse,state.annotations["${verse.bookId}.${verse.chapter}.${verse.verseStart}"],value,value?.entities?.filter {it.id in action.ids}.orEmpty())).only()
                    }
                }
                is Action.Study -> when(val delegate=(action.action as? VerseStudyFeature.Action.Delegate)?.value) {
                    is VerseStudyFeature.DelegateAction.AnnotationSaved -> state.copy(annotations=state.annotations+(delegate.value.id to delegate.value)).only()
                    is VerseStudyFeature.DelegateAction.OpenPassage -> state.copy(study=null).with(Effect.Send(Action.VisitPassage(delegate.value)))
                    VerseStudyFeature.DelegateAction.Close -> state.copy(study=null).only()
                    null -> state.only()
                }
                Action.StudyDismissed -> if(state.study?.let {it.saving || it.savedNote!=it.annotation.note}==true) state.only() else state.copy(study=null).only()
                is Action.VerseTapped -> state.with(Effect.Send(Action.StudyVerse(state.reference,action.verse)))
                Action.ClearSelectionTapped -> state.copy(selectedVerses=emptySet()).only()
                Action.CopySelectionTapped -> if(state.selectedText==null) state.only() else state.with(runEffect {clipboard.copy("${state.selectedText}\n— ${state.selectionCitation}")})
                is Action.VisitPassage -> {
                    val pos=state.positions[if(state.readingMode==ReadingMode.CONTINUOUS) "flow" else ReaderCanon.key(state.reference)] ?: Position()
                    jump(state.copy(history=(state.history+Visit(state.reference,state.flow,pos.index,pos.offset)).takeLast(32)),action.reference,bibleClient,::context)
                }
                Action.BackToReading -> state.history.lastOrNull()?.let {visit->
                    val jumped=jump(state.copy(history=state.history.dropLast(1)),visit.reference,bibleClient,::context)
                    jumped.copy(state=jumped.state.copy(flow=visit.flow,restorePosition=Position(visit.index,visit.offset)))
                } ?: state.only()
                Action.AppendChapter -> {
                    val last=state.flow.lastOrNull();val next=last?.let {ChapterNavigation.next(it)}
                    if(last==null || next==null || ReaderCanon.key(last) !in state.chapters) state.only()
                    else state.copy(flow=state.flow+next).with(Effect.Send(Action.EnsureChapter(next)))
                }
                is Action.ChapterVisible -> if(state.readingMode!=ReadingMode.CONTINUOUS || state.reference==action.reference) state.only() else state.copy(reference=action.reference,
                    content=state.chapters[ReaderCanon.key(action.reference)]?.let {Content.Loaded(it)} ?: state.content).with(Effect.Merge(listOf(context(action.reference),runEffect {preferences.setString(LAST_READ_KEY,LastRead.encode(action.reference))})))
                is Action.PositionChanged -> state.copy(positions=state.positions+(action.key to Position(action.index,action.offset))).only()
                Action.NextChapterTapped -> ChapterNavigation.next(state.reference)?.let {jump(state,it,bibleClient,::context)} ?: state.only()
                Action.PreviousChapterTapped -> ChapterNavigation.previous(state.reference)?.let {jump(state,it,bibleClient,::context)} ?: state.only()
                is Action.Go -> jump(state,action.reference,bibleClient,::context)
                is Action.TextScaleChanged -> state.copy(textScale=action.scale).only()
                is Action.ModeChanged -> state.copy(readingMode=action.mode,flow=listOf(state.reference),navigationRevision=state.navigationRevision+1,restorePosition=Position(),requestedVerses=null).with(runEffect {preferences.setString(MODE_KEY,action.mode.name)})
                is Action.FocusChanged -> state.copy(focusMode=action.enabled).with(runEffect {preferences.setString(FOCUS_KEY,action.enabled.toString())})
                Action.FocusToggled -> state.with(Effect.Send(Action.FocusChanged(!state.focusMode)))
                Action.ListenTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.Listen(state.reference))))
                Action.TalkTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.Talk(state.reference))))
                Action.ContextTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenContext(state.reference))))
                is Action.Delegate -> state.only()
            }
        },
    )
    private fun jump(state:State,reference:PassageReference,bible:BibleClient,context:(PassageReference)->Effect<Action>):Reduced<State,Action> {
        val chapter=PassageReference(reference.bookId,reference.chapter);val key=ReaderCanon.key(chapter)
        val next=state.copy(reference=chapter,requestedVerses=reference.verses,selectedVerses=emptySet(),flow=listOf(chapter),navigationRevision=state.navigationRevision+1,
            restorePosition=if(reference.verses==null && state.readingMode==ReadingMode.PAGES) state.positions[key] ?: Position() else null)
        val verses=state.chapters[key] ?: return load(next,bible)
        return next.copy(content=Content.Loaded(verses)).with(Effect.Merge(listOf(context(chapter),runEffect {send->
            ChapterNavigation.previous(chapter)?.let {send(Action.EnsureChapter(it))};ChapterNavigation.next(chapter)?.let {send(Action.EnsureChapter(it))}
        })))
    }
    private fun load(state:State,bible:BibleClient):Reduced<State,Action> {
        return state.copy(content=Content.Loading).with(Effect.Send(Action.EnsureChapter(state.reference)))
    }
}
